db/schema.go

```diff
@@ -171,23 +181,82 @@ func substituteSchemaPatch(
 	patch jsonpatch.Patch,
 	collectionsByName map[string]client.CollectionDescription,
 ) (jsonpatch.Patch, error) {
+	fieldIndexesByName := make(map[string]map[string]int, len(collectionsByName))
+	for colName, col := range collectionsByName {
+		colFieldIndexesByName := make(map[string]int, len(col.Schema.Fields))
+		fieldIndexesByName[colName] = colFieldIndexesByName
+		for i, field := range col.Schema.Fields {
+			colFieldIndexesByName[field.Name] = i
+		}
+	}
+
 	for _, patchOperation := range patch {
 		path, err := patchOperation.Path()
 		if err != nil {
 			return nil, err
 		}
 
+		path = strings.TrimPrefix(path, "/")
+		splitPath := strings.Split(path, "/")
+
 		if value, hasValue := patchOperation["value"]; hasValue {
 			var newPatchValue immutable.Option[any]
-			if isField(path) {
+			var field map[string]any
+			isField := isField(splitPath)
+
+			if isField {
 				// We unmarshal the full field-value into a map to ensure that all user
 				// specified properties are maintained.
-				var field map[string]any
 				err = json.Unmarshal(*value, &field)
 				if err != nil {
 					return nil, err
 				}
+			}
+
+			if isFieldOrInner(splitPath) {
+				fieldIndexer := splitPath[fieldIndexPathIndex]
+
+				if containsLetter(fieldIndexer) {
+					if isField {
+						if nameValue, hasName := field["Name"]; hasName {
+							if name, isString := nameValue.(string); isString && name != fieldIndexer {
+								return nil, NewErrIndexDoesNotMatchName(fieldIndexer, name)
+							}
+						} else {
+							field["Name"] = fieldIndexer
+						}
+						newPatchValue = immutable.Some[any](field)
+					}
+
+					desc := collectionsByName[splitPath[schemaNamePathIndex]]
+					var index string
+					if index == "" {
+						if colFieldIndexesByName, ok := fieldIndexesByName[desc.Name]; ok {
+							if i, ok := colFieldIndexesByName[fieldIndexer]; ok {
+								index = fmt.Sprint(i)
+							}
+						} else {
+							fieldIndexesByName[desc.Name] = map[string]int{
+								// The DocKey field is always at the zero index and we need to account for it
+								request.KeyFieldName: 0,
+							}
+						}
```

Comment:
question: If the provided collection doesn't exist, shouldn't we return an error here?

Reply:
This is me falling into the trap of coding for the future and tolerating dead code to do so.

I was concerned about when PatchSchema gains the ability to add new collections, however this block sits within an `isFieldOrInner` check, and so when the time comes it would be a poor and surprising location to do this.

I will just delete the else block.

- [ ] remove dead code

Reply:
Returning an error here would be dead code, there would be no way to reach it. If/when we allow new schema to be added via PatchSchema then we can add some logic somewhere to handle it, it would likely be located somewhere else.

Reply:
Makes sense.